zenoh/src/net/routing/hat/linkstate_peer/token.rs

```diff
@@ -0,0 +1,709 @@
+            let mut face = &mut client_tokens[0];
+            if face.whatami != WhatAmI::Client {
+                if let Some(id) = face_hat_mut!(face).local_tokens.remove(res) {
+                    face.primitives.send_declare(RoutingContext::with_expr(
+                        Declare {
+                            interest_id: None,
+                            ext_qos: ext::QoSType::DECLARE,
+                            ext_tstamp: None,
+                            ext_nodeid: ext::NodeIdType::DEFAULT,
+                            body: DeclareBody::UndeclareToken(UndeclareToken {
+                                id,
+                                ext_wire_expr: WireExprType::null(),
+                            }),
+                        },
+                        res.expr(),
+                    ));
+                }
+                for res in face_hat!(face)
+                    .local_tokens
+                    .keys()
+                    .cloned()
+                    .collect::<Vec<Arc<Resource>>>()
+                {
+                    if !res.context().matches.iter().any(|m| {
+                        m.upgrade().is_some_and(|m| {
+                            m.context.is_some()
+                                && (remote_client_tokens(&m, face)
+                                    || remote_peer_tokens(tables, &m))
+                        })
+                    }) {
+                        if let Some(id) = face_hat_mut!(&mut face).local_tokens.remove(&res) {
+                            face.primitives.send_declare(RoutingContext::with_expr(
+                                Declare {
+                                    interest_id: None,
+                                    ext_qos: ext::QoSType::DECLARE,
+                                    ext_tstamp: None,
+                                    ext_nodeid: ext::NodeIdType::DEFAULT,
+                                    body: DeclareBody::UndeclareToken(UndeclareToken {
+                                        id,
+                                        ext_wire_expr: WireExprType::null(),
+                                    }),
+                                },
+                                res.expr(),
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn forget_client_token(
+    tables: &mut Tables,
+    face: &mut Arc<FaceState>,
+    id: TokenId,
+) -> Option<Arc<Resource>> {
+    if let Some(mut res) = face_hat_mut!(face).remote_tokens.remove(&id) {
+        undeclare_client_token(tables, face, &mut res);
+        Some(res)
+    } else {
+        None
+    }
+}
+
+pub(super) fn token_remove_node(tables: &mut Tables, node: &ZenohIdProto) {
+    for mut res in hat!(tables)
+        .peer_tokens
+        .iter()
+        .filter(|res| res_hat!(res).peer_tokens.contains(node))
+        .cloned()
+        .collect::<Vec<Arc<Resource>>>()
```

Comment:
Is `collect` really needed?

Reply:
Yes. Otherwise there would be a double borrow on tables.